### ISSUE TITLE
Agent not starting due to missing jar

### ISSUE THREAD

**Issue by aravindvis** [gocd/gocd#3280]
##### Issue Type
Bug Report

##### Summary

Unable to start Go Agent after installation 17.2.0 (64 bit JRE) downloaded from the https://www.gocd.io/download/. Error summary Caused by: java.lang.RuntimeException: Agent Bootstrapper initialization failed. Could not validate file: agent-launcher.jar

##### Environment

Windows 10. 64 bit. 

Please note that the file C:\Program Files\Go Agent\agent-launcher.jar is not present in the installed folder and as a consequence the agent is not started.

###### Basic environment details

* Go Version: 17.2.0
* JAVA Version: JRE 8.0
* OS: `Windows 10`

##### Expected Results
Agent should start

##### Actual Results
Agent is not started and hence not detected in server.

##### Log snippets
<!--- 
If you could find any information/exceptions from the logs do provide it here. 
Do mask any confidential information which you don't want to be shared.
Remember to surround them in 3 backticks like this —

`
2017-03-17 12:07:22,890 [main     ] WARN  go.agent.bootstrapper.BootstrapperLoggingHelper:60 - Unable to initialize log4j file-appender: go-agent-bootstrapper.log
java.io.FileNotFoundException: go-agent-bootstrapper.log (Access is denied)
        at java.io.FileOutputStream.open0(Native Method)
        at java.io.FileOutputStream.open(Unknown Source)
        at java.io.FileOutputStream.<init>(Unknown Source)
        at java.io.FileOutputStream.<init>(Unknown Source)
        at org.apache.log4j.FileAppender.setFile(FileAppender.java:294)
        at org.apache.log4j.FileAppender.<init>(FileAppender.java:110)
        at org.apache.log4j.FileAppender.<init>(FileAppender.java:121)
        at com.thoughtworks.go.agent.bootstrapper.BootstrapperLoggingHelper.setupDefaultLog4j(BootstrapperLoggingHelper.java:54)
        at com.thoughtworks.go.agent.bootstrapper.BootstrapperLoggingHelper.initLog4j(BootstrapperLoggingHelper.java:42)
        at com.thoughtworks.go.agent.bootstrapper.AgentBootstrapper.main(AgentBootstrapper.java:56)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.lang.reflect.Method.invoke(Unknown Source)
        at com.simontuffs.onejar.Boot.run(Boot.java:306)
        at com.simontuffs.onejar.Boot.main(Boot.java:159)
2017-03-17 12:07:22,893 [main     ] WARN  go.agent.bootstrapper.BootstrapperLoggingHelper:61 - Using console-appender instead
2017-03-17 12:07:22,972 [main     ] ERROR go.agent.bootstrapper.AgentBootstrapper:39 - Error trying to validate file agent-launcher.jar: java.io.FileNotFoundException: C:\Program Files\Go Agent\agent-launcher.jar (Access is denied)
Exception in thread "main" java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.lang.reflect.Method.invoke(Unknown Source)
        at com.simontuffs.onejar.Boot.run(Boot.java:306)
        at com.simontuffs.onejar.Boot.main(Boot.java:159)
Caused by: java.lang.RuntimeException: Agent Bootstrapper initialization failed. Could not validate file: agent-launcher.jar
        at com.thoughtworks.go.agent.bootstrapper.AgentBootstrapper.validate(AgentBootstrapper.java:156)
        at com.thoughtworks.go.agent.bootstrapper.AgentBootstrapper.go(AgentBootstrapper.java:65)
        at com.thoughtworks.go.agent.bootstrapper.AgentBootstrapper.main(AgentBootstrapper.java:58)
        ... 6 more
'

**Comment by jyotisingh** [comment]
Checked on windows 2008 server box, works as expected. Have you verified the checksum of the downloaded agent? What do the logs say?

**Comment by aravindvis** [comment]
Thanks for checking. 

I did verify the checksums and they match with what is provided in the web site. I cannot find the agent-launcher.jar in the installed location. There are no errors seen on the install log as well. I am trying to get this working on my machine which is Windows 10 before I move it to my Win Server machine. Are there any restrictions on Win 10 that I might have missed?

The error I see on the logs are as below - 

Logging to go-agent-bootstrapper.log
2017-03-17 12:07:22,890 [main     ] WARN  go.agent.bootstrapper.BootstrapperLoggingHelper:60 - Unable to initialize log4j file-appender: go-agent-bootstrapper.log
java.io.FileNotFoundException: go-agent-bootstrapper.log (Access is denied)
        at java.io.FileOutputStream.open0(Native Method)
        at java.io.FileOutputStream.open(Unknown Source)
        at java.io.FileOutputStream.<init>(Unknown Source)
        at java.io.FileOutputStream.<init>(Unknown Source)
        at org.apache.log4j.FileAppender.setFile(FileAppender.java:294)
        at org.apache.log4j.FileAppender.<init>(FileAppender.java:110)
        at org.apache.log4j.FileAppender.<init>(FileAppender.java:121)
        at com.thoughtworks.go.agent.bootstrapper.BootstrapperLoggingHelper.setupDefaultLog4j(BootstrapperLoggingHelper.java:54)
        at com.thoughtworks.go.agent.bootstrapper.BootstrapperLoggingHelper.initLog4j(BootstrapperLoggingHelper.java:42)
        at com.thoughtworks.go.agent.bootstrapper.AgentBootstrapper.main(AgentBootstrapper.java:56)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.lang.reflect.Method.invoke(Unknown Source)
        at com.simontuffs.onejar.Boot.run(Boot.java:306)
        at com.simontuffs.onejar.Boot.main(Boot.java:159)
2017-03-17 12:07:22,893 [main     ] WARN  go.agent.bootstrapper.BootstrapperLoggingHelper:61 - Using console-appender instead
2017-03-17 12:07:22,972 [main     ] ERROR go.agent.bootstrapper.AgentBootstrapper:39 - Error trying to validate file agent-launcher.jar: java.io.FileNotFoundException: C:\Program Files\Go Agent\agent-launcher.jar (Access is denied)
Exception in thread "main" java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.lang.reflect.Method.invoke(Unknown Source)
        at com.simontuffs.onejar.Boot.run(Boot.java:306)
        at com.simontuffs.onejar.Boot.main(Boot.java:159)
Caused by: java.lang.RuntimeException: Agent Bootstrapper initialization failed. Could not validate file: agent-launcher.jar
        at com.thoughtworks.go.agent.bootstrapper.AgentBootstrapper.validate(AgentBootstrapper.java:156)
        at com.thoughtworks.go.agent.bootstrapper.AgentBootstrapper.go(AgentBootstrapper.java:65)
        at com.thoughtworks.go.agent.bootstrapper.AgentBootstrapper.main(AgentBootstrapper.java:58)
        ... 6 more

**Comment by aravindvis** [comment]
Hi,
I wasn't able to start the server from Windows Services menu and hence had reverted to the command prompt when I saw the error for which I had placed the stacktrace above. Looking at the error message of access denied for one of the earlier jars, I ran the command prompt under Admin mode and with that saw a different error - 


C:\Program Files\Go Agent>java -jar agent-bootstrapper.jar  -serverUrl
logFile Environment Variable= null
Logging to go-agent-bootstrapper.log
Expected a value after parameter -serverUrl
Usage: java -jar agent-bootstrapper.jar [options]
  Options:
    -help
      Print this help
      Default: false
    -rootCertFile
      The root certificate from the certificate chain of the GoCD server (in
      PEM format)
  * -serverUrl
      The GoCD server URL. Must begin with `https://`, and end with `/go`
    -sslVerificationMode
      The SSL verification mode.
      Default: NONE
      Possible Values: [FULL, NONE, NO_VERIFY_HOST]

Realized this is the error I see on the logs when I run from the services menu as well. However, when I start the server under Admin Mode with the server url being passed as a param 

java -jar agent-bootstrapper.jar  -serverUrl https://localhost:8154/go

The agent starts up and I am able to see under the agents tab. 

Thanks,
Aravind
